### PR TITLE
Implemented new Group handshake extension format

### LINKED ISSUE #1245: Change the Group Membership extension of the handshake packet
The Group Membership extension of the SRT handshake packet is used to exchange group connection configuration and establish a bonded SRT connection. The functionality to be introduced in v1.5.0 release.

In the current implementation, SRT receiver in "broadcast" and "backup" modes matches packets by the "Packet Sequence Number" field of the SRT Data packet.
In the "balancing" mode the receiver arranges incoming packets based on the Message Number field.
The used mode of arrangement should be signaled in the corresponding handshake extension to allow future enhancements.

### Expected Format

The expected structure of the Group Membership extension is the following:

```
0                   1                   2                   3
0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                           Group ID                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
| Group Type  | Group's Flags |       Group's Weight            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

where Group's Flags has the following structure:
```
0 1 2 3 4 5 6 7 
+-+-+-+-+-+-+-+
|             |
+-+-+-+-+-+-+-+
|   (zero)  |M|
+-+-+-+-+-+-+-+
```
with M = 1 for synchronization based on message number. 0 - for arrangement based on packet sequence numbers.

### Current Format

As of now, the extension has the following structure:
```
0                   1                   2                   3
0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                           Group ID                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                         Group Type                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                       Group's Weight                          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

### PR DESCRIPTION
**Pull request by ethouris**
Fixes #1245

Applied changes:

1. The format of the data in the group extension is according to #1245. 
2. The flag deciding about synchronizing packets by msgno is added. This flag solely - not the group type, as tried before - decides whether synchronization goes on message numbers or sequence numbers.
3. Also added the condition of using synchronized sequence numbers only when this flag is false.
4. Also recognizing this flag in the handshake and sending the "fixed" value of the flag in case when there were a newer caller, which supports also balancing groups with sequence synchronization, is used with older listener that doesn't.

### CODE REVIEW

File: srtcore/handshake.h [hunk 1]
```diff
@@ -150,6 +149,18 @@ struct SrtHSRequest: public SrtHandshakeExtension
         return base | SRT_HSTYPE_ENCFLAGS::wrap( SRT_PBKEYLEN_BITS::unwrap(crypto_keylen) );
     }
 
+    // Group handshake extension layout
+//  0                   1                   2                   3
+//  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//  |                           Group ID                            |
+//  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//  | Group Type  | Group's Flags |       Group's Weight            |
```
Comment (maxsharabayko) [inline srtcore/handshake.h:158]: The alignment of the first line does not match the alignment of the following lines. Intended?

File: srtcore/core.h [hunk 2]
```diff
@@ -134,11 +134,9 @@ extern const SRT_SOCKOPT srt_post_opt_list [];
 enum GroupDataItem
 {
     GRPD_GROUPID,
-    GRPD_GROUPTYPE,
-    GRPD_WEIGHT,
+    GRPD_GROUPDATA,
 
-
-    GRPD__SIZE
+    GRPD_E_SIZE
```
Comment (maxsharabayko) [inline srtcore/core.h:139]: Why is the weight deleted from here?

File: srtcore/core.cpp [hunk 3]
```diff
@@ -1797,7 +1805,14 @@ size_t CUDT::fillHsExtGroup(uint32_t* pcmdspec)
...
 
-    int32_t storedata [GRPD__SIZE] = { id, tp, m_parent->m_IncludedIter->weight };
+    // See CUDT::interpretGroup()
+
+    uint32_t dataword = 0
+        | SrtHSRequest::HS_GROUP_TYPE::wrap(tp)
+        | SrtHSRequest::HS_GROUP_FLAGS::wrap(flags)
+        | SrtHSRequest::HS_GROUP_WEIGHT::wrap(m_parent->m_IncludedIter->weight);
+
+    uint32_t storedata [GRPD_E_SIZE] = { uint32_t(id), dataword };
```
Comment (maxsharabayko) [inline srtcore/core.cpp:1815]: const?

File: srtcore/core.h (hunk 2 above)
Comment (ethouris) [reply srtcore/core.h:139]: These are indexes to the `uint32_t` array. Previously it was type and weight packet to separate cells. Now all - weight, type and flags - are in a single cell, therefore it uses the index name collectively as "group data".

Comment (maxsharabayko) [reply srtcore/core.h:139]: Ok, makes sense.

**Pull request merged**